ni_measurementlink_service/session_management/_reservation.py

```diff
@@ -183,6 +285,90 @@ def _create_sessions_core(
+        results = self._get_connections_core(
+            session_type, pin_or_relay_names, sites, instrument_type_id
+        )
+        if not results:
+            raise ValueError(
+                "No reserved connections matched the specified criteria. "
+                "Expected single connection, got 0 connections."
+            )
+        elif len(results) > 1:
+            raise ValueError(
+                "Too many reserved connections matched the specified criteria. "
+                f"Expected single connection, got {len(results)} connections."
+            )
+
+        return results[0]
+
+    def _get_connections_core(
+        self,
+        session_type: Type[TSession],
+        pin_or_relay_names: Union[str, Iterable[str], None] = None,
+        sites: Union[int, Iterable[int], None] = None,
+        instrument_type_id: Optional[str] = None,
+    ) -> Sequence[TypedConnection[TSession]]:
+        requested_pins = _to_iterable(pin_or_relay_names, self._reserved_pin_or_relay_names)
+
+        if sites == SITE_ALL_SITES:
+            requested_sites = self._reserved_sites
+        else:
+            requested_sites = _to_iterable(sites, self._reserved_sites)
+
+        requested_sites_with_system = requested_sites
+        if SITE_SYSTEM_PINS not in requested_sites_with_system:
+            requested_sites_with_system = list(requested_sites_with_system)
+            requested_sites_with_system.append(SITE_SYSTEM_PINS)
+
+        requested_instrument_type_ids = _to_iterable(
+            instrument_type_id, self._reserved_instrument_type_ids
+        )
+
+        # Sort the results by site, then by pin, then by instrument type (as a tiebreaker).
+        results: List[TypedConnection[TSession]] = []
+        matching_pins: Set[str] = set()
+        matching_sites: Set[int] = set()
+        for site in requested_sites_with_system:
+            for pin in requested_pins:
+                for instrument_type in requested_instrument_type_ids:
+                    key = _ConnectionKey(pin, site, instrument_type)
+                    value = self._connection_cache.get(key)
+                    if value is not None:
+                        session = self._session_cache.get(value.session_info.session_name)
+                        value = value._with_session(session)
+                        value._check_runtime_type(session_type)
+                        results.append(cast(TypedConnection[TSession], value))
+                        matching_pins.add(pin)
+                        matching_sites.add(site)
+
+        if pin_or_relay_names is not None:
+            missing_pins = set(requested_pins) - matching_pins
+            if missing_pins:
+                missing_pin_list = ", ".join(f"'{pin}'" for pin in sorted(missing_pins))
+                raise ValueError(
+                    f"No reserved connections matched pin or relay name(s) {missing_pin_list}."
+                )
+
+        if sites is not None:
```

Comment:
We don't do this check in LabVIEW. My main concern is in that in the all sites case, do you really want an error if you don't get a connection for at least every site? Maybe the answer is yes. I'm not sure. It seemed to heavy handed to me.

Reply:
In general, all sites should have the same pins. If you create a site that doesn't have connections for all of the DUT pins, the pin map editor displays a warning:

> Warning: An error might occur at run-time.
> The following pins in the pin map are not connected:
>
> DUTPin1 - Sites: 1

I could change the code to check for missing sites only when there are no results, but this will allow passing a non-existent site like `sites=[0, 1, 3873875837]`, as long as one of the other sites matches a connection.

I could change the code to check whether the requested sites are a subset of the reserved sites up-front. Perhaps this is better.

Note that this code only knows which pins and sites are reserved, not which pins and sites exist in the pin map. This means that the `SITE_ALL_SITES` constant actually matches against all reserved sites, not all sites. Perhaps I should rename it?

Cc: @jonathanmendez @jasonmreding

Reply:
I don't know that I have a strong opinion either way. For the most part, I think it won't really matter or have much impact on our users. There are perhaps some corner cases involving shared pins or system pins where this could cause an issue, but I suspect in most cases even that won't be an issue. 

I tend to view sites and instrument type ID inputs as equivalent in regards to this function so I would skew to keeping their behavior the same. So if one errors then I would also error for the other. If one doesn't throw an error, then I wouldn't throw an error for the other one either. In most cases, I suspect only throwing an error if there are no matching pins is sufficient and would cover no matching sites or instrument type ID as well. Is there any case where returning an empty list of connections without throwing an exception would be acceptable/preferable? Is the intent to also use this API to get at the "raw" session info without performing any filtering? If so, would it make sense to throw an error whenever there are no matching pins when any of the pins, instrument type ID, or sites were explicitly set? Or was this the intent all along, and we are just trying to provide a more targeted, helpful error message?

Reply:
I want to avoid garbage-in, garbage out. If the user misspells a pin name or instrument type ID, I would prefer to return an error rather than silently ignore it. So I am adding pre-query checks to verify that the requested pins, sites, and instrument type ID match *something* in isolation (even if they do not match anything when combined).

This leaves two cases for post-query checks:
- You specified pin names, but we didn't find anything that matched them. If you got past the pre-query check, the pin names are valid, so this condition must mean that the combination of pin name and other criteria (site and instrument type ID) didn't match anything. So I think this could throw a slightly more descriptive error.
- For whatever reason, nothing matched. Perhaps you used reserve_sessions(), nothing was reserved, and you didn't specify any matching criteria. I guess it's okay for get_connections() to return an empty list here. get_connection() will still error.

> In most cases, I suspect only throwing an error if there are no matching pins is sufficient and would cover no matching sites or instrument type ID as well.

With the added pre-query checks, I agree. Any post-query check failure means that the combination of criteria failed to match.

> Is there any case where returning an empty list of connections without throwing an exception would be acceptable/preferable? 

Maybe the case where reserve_sessions() reserved nothing.

> Is the intent to also use this API to get at the "raw" session info without performing any filtering?

Yes, it can return all reserved connections.

> If so, would it make sense to throw an error whenever there are no matching pins when any of the pins, instrument type ID, or sites were explicitly set?

That seems plausible.

Reply:
I updated the error handling as described in my last reply.

Reply:
What you have now seems reasonable. I will create a work item so that LabVIEW can match this behavior.